h2/src/main/org/h2/bnf/context/DbContextRule.java

```diff
@@ -144,16 +147,16 @@ public boolean autoComplete(Sentence sentence) {
             if (query.indexOf(' ') < 0) {
                 break;
             }
-            for (; i < up.length(); i++) {
-                char ch = up.charAt(i);
+            for (; i < query.length(); i++) {
+                char ch = query.charAt(i);
                 if (ch != '_' && !Character.isLetterOrDigit(ch)) {
                     break;
                 }
             }
```

Comment:
This old code can't read all identifiers properly. It should be
if (i >= l || ((cp = query.codePointAt(i)) != '_' && !Character.isJavaIdentifierStart(cp))) {
    break;
}
while ((i += Character.charCount(cp)) < l && Character.isJavaIdentifierPart(cp = query.codePointAt(i))) {
    //
}
```Java
int l = query.length(), cp;
if (i >= l || ((cp = query.codePointAt(i)) != '_' && !Character.isJavaIdentifierStart(cp))) {
    break;
}
while ((i += Character.charCount(cp)) < l && Character.isJavaIdentifierPart(cp = query.codePointAt(i))) {
    //
}
```
In some compatibility modes of H2 and in some other DBMS `#` is valid too, but I think we can ignore this deviation here.